src/Storages/StorageMerge.cpp

```diff
@@ -1020,33 +1011,42 @@ StorageMerge::StorageListWithLocks StorageMerge::getSelectedTables(
         }
     }
 
+    if (!filter_by_database_virtual_column && !filter_by_table_virtual_column)
+        return res;
+
+    auto filter_actions_dag = ActionsDAG::buildFilterActionsDAG(filter_nodes.nodes, {}, context);
+    if (!filter_actions_dag)
+        return res;
+
+    const auto * predicate = filter_actions_dag->getOutputs().at(0);
+
     if (filter_by_database_virtual_column)
     {
         /// Filter names of selected tables if there is a condition on "_database" virtual column in WHERE clause
         Block virtual_columns_block
             = Block{ColumnWithTypeAndName(std::move(database_name_virtual_column), std::make_shared<DataTypeString>(), "_database")};
-        VirtualColumnUtils::filterBlockWithQuery(query, virtual_columns_block, query_context);
+        VirtualColumnUtils::filterBlockWithPredicate(predicate, virtual_columns_block, query_context);
         auto values = VirtualColumnUtils::extractSingleValueFromBlock<String>(virtual_columns_block, "_database");
 
         /// Remove unused databases from the list
-        selected_tables.remove_if([&](const auto & elem) { return values.find(std::get<0>(elem)) == values.end(); });
+        res.remove_if([&](const auto & elem) { return values.find(std::get<0>(elem)) == values.end(); });
     }
 
     if (filter_by_table_virtual_column)
     {
         /// Filter names of selected tables if there is a condition on "_table" virtual column in WHERE clause
         Block virtual_columns_block = Block{ColumnWithTypeAndName(std::move(table_name_virtual_column), std::make_shared<DataTypeString>(), "_table")};
-        VirtualColumnUtils::filterBlockWithQuery(query, virtual_columns_block, query_context);
+        VirtualColumnUtils::filterBlockWithPredicate(predicate, virtual_columns_block, query_context);
```

Comment:
This is the reason for refactoring.
More or less, everything else is code movement.